ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -316,6 +314,30 @@ private boolean removeFiles(String location, ValidWriteIdList writeIdList, Compa
       }
       fs.delete(dead, true);
     }
-    return true;
+    // Check if there will be more obsolete directories to clean when possible. We will only mark cleaned when this
+    // number reaches 0.
+    return getNumEventuallyObsoleteDirs(location, dirSnapshots) == 0;
+  }
+
+  /**
+   * Get the number of base/delta directories the Cleaner should remove eventually. If we check this after cleaning
+   * we can see if the Cleaner has further work to do in this table/partition directory that it hasn't been able to
+   * finish, e.g. because of an open transaction at the time of compaction.
+   * We do this by assuming that there are no open transactions anywhere and then calling getAcidState. If there are
+   * obsolete directories, then the Cleaner has more work to do.
+   * @param location location of table
+   * @return number of dirs left for the cleaner to clean – eventually
+   * @throws IOException
+   */
+  private int getNumEventuallyObsoleteDirs(String location, Map<Path, AcidUtils.HdfsDirSnapshot> dirSnapshots)
+      throws IOException {
+    ValidTxnList validTxnList = new ValidReadTxnList();
+    //save it so that getAcidState() sees it
+    conf.set(ValidTxnList.VALID_TXNS_KEY, validTxnList.writeToString());
+    ValidReaderWriteIdList validWriteIdList = new ValidReaderWriteIdList();
+    Path locPath = new Path(location);
+    AcidUtils.Directory dir = AcidUtils.getAcidState(locPath.getFileSystem(conf), locPath, conf, validWriteIdList,
+        Ref.from(false), false, dirSnapshots);
+    return dir.getObsolete().size();
```

Comment:
should we consider aborts as well?

Reply:
You should not, there can be aborts with higher txnId than compaction, this will see those, but the cleaner will never see them, so it would never finish its job

Reply:
there could be deltas with higher txnId than compaction :) 
If we wan't handle this, we might get uncleaned aborts

Reply:
Aborted directories are deleted by removeFiles() too, and there can be obsolete directories with a higher txnId than this compaction too. The point of this change is to call markCleaned() when the table is completely cleaned, not necessarily when this compaction's cleaning is done. I figured that would be better than calling markCleaned() when some obsolete or aborted file gets deleted. (see HIVE-24444 description)

Reply:
I think checking the aborted is a very problematic in upstream, if you have transaction coming in continuously and some of them aborting cleaner won't be able to mark any compaction cleaned. Example:
txn1-10 coming in some of them writing, txnid=8 is aborting
compaction is kicking in, compacts the deltas, will have the next_txn_id for example txnId=13
other transactions are coming in txnid=18 is aborting
cleaner will start when everything is done under 13, it can clean everything. but since txnId=14 is open it will not see txnId=18, so when it checks if can mark itself cleaned, it will not do it, because the check sees txnId=18 aborted.
Second compaction kicks in, will have next_txn_id for example txnId=23
other transactions are coming in txnid=28 is aborting.
Again cleaner can not mark compaction 1 or compaction 2 cleaned, as none of them see 28 and can not clean it. this will pile up and Cleaner only mark them cleaned, if finally there is an iteration, when there was no new aborted transaction.
If you check only the obsoleted files, you only miss marking the compaction cleaned, when you should have, if you have two overlapping compaction, but in that case the second compaction will get marked as cleaned and the next cleaner run, compaction 1 will be marked as cleaned too. This too can cause problems if you enabled delayed cleaner, but probably less

Reply:
Okay, I see what you mean. I removed the aborted files from the total.

In general do you think checking for obsolete files is better than checking whether we removed any files?
Case 1: Assuming HIVE-23107 etc. are present in the version?
Case 2: Assuming HIVE-23107 etc. are _not_ present in the version?

Reply:
Case 1: If HIVE-23107 and the following are there I think none of these checks are necessary, because we can be sure, that Cleaner was running when it could delete everything it can. Also if delayed cleaning is enabled it is guaranteed, that it will never delete any more obsolete directories no matter how many times it is running (see: validWriteIdList.updateHighWatermark(ci.highestWriteId)). If we must choose, checking if anything was removed does less damage
Case 2: If those fixes are not there I think checking for obsolete files is better than checking if anything was removed

Reply:
considering Case2:
How the situation is going to change if instead of aborted txn you have successful write? Writes coming in continuously (streaming). Won't new delta prevent us from marking clean as complete for corresponding compaction request and we'll have just pilled up cleanup requests in a queue?

Reply:
New deltas by themselves don't stop marking the compaction cleaned (they are not in obsolete list), but overlapping compactions do, I think problematic scenario could be if there are some long running ETL jobs also next to your streaming writes. That would mean the that cleaning jobs would pile up, they would run continuously, but the low values in min_history_level would prevent them to clean every obsolete up and ever marked as cleaned :(

Reply:
Why new deltas won't appear in obsolete list (getNumEventuallyObsoleteDirs) if we consider every transaction as valid?

Reply:
It becomes obsolete, when there is a base / minor compacted delta shadowing it. Otherwise they are just in currentDirectories list.

Reply:
Why don't we keep only 1 clean request per table/partition and just replace existing one?

Reply:
I guess that's doable...and would fix the issue at hand.

The cons I can think of:
It might slow down the worker a bit, since every time a table/partition goes into "ready for cleaning" state it needs to go through the whole compaction queue. And it might confuse users since entries will be deleted from the compaction queue.

Reply:
In upstream it would cause problems, for the delayed cleaner it is necessary that every compaction entry cleans only its own obsoletes. The default is every compaction is delayed by 15 minutes, so it is much more likely when a cleaning job finally runs, there are already some other jobs in the queue waiting. all of them contains highestwriteId saved, and they will clean only up until that. You should not merge those.

Reply:
you can merge stuff below retention period in this case

Reply:
I haven't checked, but if folder is considered as obsolete when there is a base / minor compacted delta shadowing it - that's ok. 
But i don't see benefits of this patch for upstream (contains HIVE-23107), am I missing something?

Reply:
No, this isn't necessary upstream, this change is for versions without HIVE-23107 etc. But I don't want to hurt upstream functionality with it.

Reply:
if it's only for versions without HIVE-23107, should it be behind the feature flag/schema version check?